File: src/main/java/architecture/goldenboughs_lib/config/LibConfigUtil.java
package architecture.goldenboughs_lib.config;

import architecture.goldenboughs_lib.core.GoldenBoughsLib;
import net.neoforged.neoforge.common.ModConfigSpec;
import org.apache.commons.lang3.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import java.util.function.Function;

public abstract class LibConfigUtil {
	protected static ModConfigSpec.@NotNull DoubleValue define(
		ModConfigSpec.@NotNull Builder builder,
		double defaultValue, double min, double max,
		String key,
		String... comment
	) {
		return builder.comment(comment)
			.translation(getTranslation(key))
			.defineInRange(key, defaultValue, min, max);
	}

	public static @NotNull String getTranslation(String @NotNull ... keys) {
		if (keys.length == 0) {
			return GoldenBoughsLib.ID + ".config";
		}
		StringBuilder builder = new StringBuilder();
		for (String key : keys) {
			builder.append(".");
			builder.append(key);
		}
		return GoldenBoughsLib.ID + ".config" + builder;
	}

	protected static ModConfigSpec.@NotNull IntValue define(
		ModConfigSpec.@NotNull Builder builder,
		int defaultValue, int min, int max,
		String key,
		String... comment
	) {
		return builder.comment(comment)
			.translation(getTranslation(key))
			.defineInRange(key, defaultValue, min, max);
	}

	protected static ModConfigSpec.@NotNull BooleanValue define(
		ModConfigSpec.@NotNull Builder builder,
		boolean defaultValue,
		String key,
		String... comment
	) {
		return builder.comment(comment)
			.translation(getTranslation(key))
			.define(key, defaultValue);
	}

	protected static <T> @NotNull Pair<T, ModConfigSpec> configure(Function<ModConfigSpec.Builder, T> consumer) {
		return new ModConfigSpec.Builder().configure(consumer);
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/config/LibConfigUtil.kt
package architecture.goldenboughs_lib.config

import architecture.goldenboughs_lib.core.Lib
import net.neoforged.neoforge.common.ModConfigSpec
import org.apache.commons.lang3.tuple.Pair
import java.util.function.Function
import architecture.goldenboughs_lib.core.LibConstants

abstract class LibConfigUtil {
	companion object {
		@JvmStatic
		protected fun ModConfigSpec.Builder.define(
			defaultValue: Double,
			min: Double,
			max: Double,
			key: String,
			vararg comment: String
		): ModConfigSpec.DoubleValue {
			return comment(*comment)
				.translation(getTranslation(key))
				.defineInRange(key, defaultValue, min, max)
		}

		@JvmStatic
		fun getTranslation(vararg keys: String): String {
			if (keys.isEmpty()) {
				return LibConstants.ID + ".config"
			}
			val builder = StringBuilder()
			for (key in keys) {
				builder.append(".")
				builder.append(key)
			}
			return LibConstants.ID + ".config" + builder
		}

		protected fun ModConfigSpec.Builder.define(
			defaultValue: Int,
			min: Int,
			max: Int,
			key: String,
			vararg comment: String
		): ModConfigSpec.IntValue {
			return comment(*comment)
				.translation(getTranslation(key))
				.defineInRange(key, defaultValue, min, max)
		}

		@JvmStatic
		protected fun ModConfigSpec.Builder.define(
			defaultValue: Boolean,
			key: String,
			vararg comment: String
		): ModConfigSpec.BooleanValue {
			return comment(*comment)
				.translation(getTranslation(key))
				.define(key, defaultValue)
		}

		@JvmStatic
		protected fun <T> configure(consumer: Function<ModConfigSpec.Builder, T>): Pair<T, ModConfigSpec> {
			return ModConfigSpec.Builder().configure<T>(consumer)
		}
	}
}
